extensions/data-transfer/portability-data-transfer-microsoft/src/main/java/org/dataportabilityproject/transfer/microsoft/transformer/calendar/ToCalendarEventModelTransformer.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2018 The Data-Portability Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.dataportabilityproject.transfer.microsoft.transformer.calendar;
+
+import org.dataportabilityproject.transfer.microsoft.transformer.TransformerContext;
+import org.dataportabilityproject.transfer.microsoft.transformer.common.TransformerHelper;
+import org.dataportabilityproject.types.transfer.models.calendar.CalendarAttendeeModel;
+import org.dataportabilityproject.types.transfer.models.calendar.CalendarEventModel;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiFunction;
+
+import static java.util.Collections.emptyList;
+import static org.dataportabilityproject.transfer.microsoft.transformer.TransformConstants.CALENDAR_ID;
+
+/**
+ * Maps from a Graph API calendar event resource as defined by: https://developer.microsoft.com/en-us/graph/docs/api-reference/beta/resources/event
+ */
+public class ToCalendarEventModelTransformer implements BiFunction<Map<String, Object>, TransformerContext, CalendarEventModel> {
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public CalendarEventModel apply(Map<String, Object> event, TransformerContext context) {
+        if (!"singleInstance".equals(event.get("type"))) {   // support single instances for now;recurring events later
+            return null;
+        }
+        String calendarId = context.getProperty(CALENDAR_ID);
```

Comment:
Do we always know we have a calendar_id in the context by this point?

Reply:
Yes. I think this will come up for other transformations as well that require some form of context data to complete a transformation. I don't think we want to bake that into a context signature since this type of data can vary widely.

Reply:
Agreed, I'm just wondering what if we get an event without the outer calendar information, do we know the ID?

Reply:
Ah OK. In the MS API the exporter always exports Calendar containers first.